man/man1/fitstab.1
.TH FITSTAB 1NEMO "10 November 2005"

.SH "NAME"
fitstab \- convert fits table to ascii table

.SH "SYNOPSIS"
.PP
\fBfitstab in=\fPfits_table [parameter=value]

.SH "DESCRIPTION"
\fBfitstab\fP reads a fits table file from disk and converts it to
an ascii table. The input file must be in \fIfits(5NEMO)\fP format,  
the output will be on \fIstdout\fP. Tables can be selected by their
HDU number; HDU's which do not have the ascii table extension (TABLE) 
are automatically skipped.
.PP
See notes in \fIccdfits(1NEMO)\fP how to process fits files from
tape instead disk.

.SH "PARAMETERS"
The following parameters are recognized in any order if the keyword is also
given:
.TP 20
\fBin=\fIfits-file\fP
input file, must be \fIfits(5)\fP format.
.br
[no default].
.TP
\fBhdu=\fInumber\fP
HDU number to be processed. The first one counts as 1. A 0 will
convert all. 
HDU's which do not have the table extension 
(\fBXTENSION='TABLE     '\fP) are automatically skipped.
.br
[Default: 0].
.TP
\fBselect=\fP
Comma separated list of 
selections which determines the parts of the table that is output.
Valid options are \fBdata\fP, \fBheader\fP and \fBrow\fP:
\fBdata\fP will list the table data itself, with an added
row number if the option \fBrow\fP is also supplied, and
\fBheader\fP lists the column names and their attributes from
the FITS header. The parameters from a random group HDU can also
be printed using \fBgroup\fP.
.br
[Default: \fBdata\fP].
.TP
\fBcol=\fIcolumn_names\fP
Column (or random group parameter)
names selected for output. Default is all columns, in the original
format (no extra blanks). If columns are selected by their name
(FITS keyword \fBTTYPE\fPn), the output
will contains an extra blank inbetween columns, UNLESS that column was
followed by a \fB+\fP sign, for example \fBcol=DecSign+DecD\fP will not
put an extra space between \fBDecSign\fP and \fBDecD\fP. \fI*** this last
item not implemented yet ***\fP
.br
[Default: all].
.TP
\fBfnl=\fP
Frequency of newlines in the output, meaning every \fBfnl\fP numbers that
are written, a newline will be added to the output.  Use 0 if no extra
newlines are needed.
.br
[Default: 0]

.SH "BUGS"
Columns are output in the order that they appear in the table, not
as specified with the \fBcol=\fP keyword.
.PP
no magic blanking substitution, no scaling TSCALnnn, TZEROnnn ?
.PP
column names are treated case sensitive. 
.PP
X format not understood yet

.SH "EXAMPLES"
Show the random group parameters of the first 4 record:
.nf
% fitstab sim0.fts hdu=1 row=1:4 select=group,data
Working on FITS file 1
Random Group printing:
RANDOM GROUP: 4950 rows with 10 (4-byte) length each; 7 pars
# UU VV WW DATE DATE BASELINE FREQSEL Data(1..)
 -8.12176e-07 1.32065e-07 0 2.47447e+06 0.572593 258 1 858.081 199.085 7.5
 5.68985e-08 -4.70039e-07 0 2.47447e+06 0.572593 259 1 1251.84 -22.7329 7.5
 -3.45473e-07 6.32915e-07 0 2.47447e+06 0.572593 260 1 893.396 27.1245 7.5
 -8.36404e-07 -5.8816e-07 0 2.47447e+06 0.572593 261 1 518.19 129.551 7.5
.fi

.SH "SEE ALSO"
ccdfits(1NEMO), fits(5NEMO), dd(1), FTB (\fIADC Fits Table Browser\fP), fv(1ftools)

.SH "AUTHOR"
Peter Teuben

.SH "FILES"
.nf
.ta +2.5i
~/src/image/fits  	fitstab.c
.fi

.SH "HISTORY"
.nf
.ta +1.0i +4.0i
8-apr-92	V1.0 Written	PJT
13-apr-92	V1.0x Renamed file= to hdu=	PJT
30-sep-96	V1.3 added output of random group parameters	PJT
10-nov-05	V1.5 added fnl=	PJT
.fi



src/image/fits/fitstab.c
/*
 *   FITSTAB:   output fits ASCII tables (XTENSION='TABLE   ')
 *
 *	8-apr-92    V1.0  Written       PJT
 *     15-apr-92    V1.1  added row=    PJT
 *     24-apr-92        a fixed bug in row= introduced earlier  PJT
 *     18-jan-92    V1.2  attempt to handle BINTABLE too  (fits.c)     PJT
 *     30-sep-96    V1.3  option to print out random group pars        pjt
 *
 *     10-nov-05    V1.5  add newline=
 * BUT PERHAPS WE SHOULD not boolean it, but with an integer give how often newlines are written
 *    
 */

#include <stdinc.h>
#include <getparam.h>
#include <fits.h>

string defv[] = {			/* Standard NEMO keyword+help */
    "in=???\n              Input fits table file",
    "hdu=0\n               Select which HDU file(s)? [0=all]",
    "select=data\n         Output mode {info,data,header,row,group}",
    "col=\n                Column names to select [all]",
    "row=\n                Row numbers to select [all]",
    "format=\n             Format for each row to use [TFORMnnn]",
    "maxrow=20000\n        Max row numbers to select",
    "random=f\n            Force random group?",
    "fnl=0\n               Frequency of newlines",
    "VERSION=1.5a\n        13-jul-2021 PJT",
    NULL,
};

string usage = "convert fits table or random groups to ascii table";

//extern string *burststring(string,string);

void nemo_main()
{
    stream instr;
    off_t  n;
    int    i, nfile, maxrow, nrow, *row=NULL;
    string *col, select;
    struct fits_header fh;
    bool   scanopt();
    bool   Qrg = getbparam("random");
    int    fnl = getiparam("fnl");

    instr = stropen(getparam("in"),"r");    /* open input */
    
    if (hasvalue("col"))		/* select some columns */
        col = burststring(getparam("col"),", ");
    else				/* or select them all */
        col = NULL;
    select = getparam("select");
    nfile = getiparam("hdu");		/* need: nemoinpi() */
    maxrow = getiparam("maxrow");
    if (hasvalue("row")) {
        row = (int *) allocate((maxrow+1)*sizeof(int));
        nrow = nemoinpi(getparam("row"),row,maxrow);
        /* sorti(row,nrow); */
        if (nrow<0) {
            warning("Cannot use row selection - will use all");
            row = NULL;
        } else
            row[nrow] = 0;      /* signal end of list */
    }

    for (i=1;;i++) {			             /* try infinite loop */
       fts_zero(&fh);			             /* clean out header */
       n = fts_rhead(&fh,instr);	               /* read header */
       if (n<0)				              /* if no data (EOF) .. */
          break;			              /* ... quit */
       dprintf(1,"Working on FITS file %d\n",i);

       if ((nfile==0 || nfile==i)) {			/* if right file */
	 if (Qrg)
	   fts_pgroup(&fh,instr,col,select,row,fnl); 	/* try and print */
	 else
	   fts_ptable(&fh,instr,col,select,row,fnl); 	/* try and print */
       } else                                            /* Otherwise just */
	 fts_sdata(&fh,instr);		               /* skip the data */

       if (i==nfile)
            break;                                        /* done */
    }
    if (nfile>0 && i>=nfile && n<0)
        stop(1);                    /* see if early eof */

    strclose(instr);                    /* close files */
}
